### PR TITLE
feat(kic 2.0) add profiling

### LINKED ISSUE #1298: KIC 2.0: Handle profiling flag
1.x uses the `--profiling` flag to enable [pprof](https://golang.org/pkg/net/http/pprof/) endpoints [on the health/metrics/etc. HTTP server](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.2.0/cli/ingress-controller/main.go#L698-L709).

2.x uses the controller-runtime default servers, so it doesn't immediately have a place to put this. We'd need to add a custom server to still expose profiling data. The rest of the implementation is unlikely to change, since we just call pprof's functions.

### PR DESCRIPTION
**Pull request by tharun208**
Signed-off-by: Tharun <rajendrantharun@live.com>

**What this PR does / why we need it**: This PR introduces profiling for kic 2.x

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1298 

**Special notes for your reviewer**:
 Created a separate server as when #1278 is implemented it will be easy to add to this change.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by tharun208** [comment]
raised a new pr as #1408 messed up. either we can go with this one or #1416

File: railgun/internal/diagnostics/server.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	if enableProfiling {
+		mgrutils.Install(mux)
+	}
+	return &diagnosticsServer{
+		logger: log,
+		mux:    mux,
+	}
+}
+
+func (s *diagnosticsServer) Start(ctx context.Context) error {
```
Comment (mflendrich) [inline railgun/internal/diagnostics/server.go:29]: This function blocks for the lifetime of the profiling server, while `Start` implies that this function just "starts" the server (in the background).

Please rename to `Listen`.

File: railgun/internal/diagnostics/server.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/kong/kubernetes-ingress-controller/railgun/pkg/config"
+	"github.com/sirupsen/logrus"
+)
+
+type diagnosticsServer struct {
+	logger logrus.FieldLogger
+	mux    *http.ServeMux
+}
+
+func NewDiagnosticsServer(enableProfiling bool, log logrus.FieldLogger) *diagnosticsServer {
```
Comment (mflendrich) [inline railgun/internal/diagnostics/server.go:18]: To simplify the implementation, we can remove `NewDiagnosticsServer` altogether:
- remove the `mux` field from the struct and make it a local variable in `Start()` and initialize it there
- remove `NewDiagnosticsServer` and, with the comment above (which makes the type exported) addressed, export the `Logger` field of `Server`

Therefore the initialization becomes:
```go
s := diagnostics.Server{Logger: util.MakeLogger(c.LogLevel, c.LogFormat)}
if err := s.Listen(ctx); err != nil { ... } // renamed in another comment
```

**Review by mflendrich — CHANGES_REQUESTED** [review]
Some refactoring requests attached - the implementation and the test look neat otherwise!

File: CHANGELOG.md [hunk 3]
```diff
@@ -36,6 +36,8 @@ released and the release notes may change significantly before then.
 
 #### Added
 
+- added support for profiling in KIC 2.0.
```
[the remark this answers is not in the document]
Comment (tharun208) [reply CHANGELOG.md:39]: addressed

File: railgun/internal/mgrutils/profiling.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package mgrutils
+
+import (
+	"net/http"
+	"net/http/pprof"
+)
+
+// Install adds the Profiling webservice to the given mux.
+func Install(mux *http.ServeMux) {
```
[the remark this answers is not in the document]
Comment (tharun208) [reply railgun/internal/mgrutils/profiling.go]: this is under a separate pkg. can I move this to `railgun/internal/diagnostics/server.go ` ?

[the remark this answers is not in the document]
Comment (mflendrich) [reply railgun/internal/mgrutils/profiling.go]: Yes of course. I should have made that more clear.

[the remark this answers is not in the document]
Comment (tharun208) [reply railgun/internal/mgrutils/profiling.go]: addressed

**Review by mflendrich — APPROVED** [review]
Looks good!

**Comment by rainest** [comment]
@tharun208 Thank you for your contribution. Please fill out the following form to claim your contributor swag:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributor-t-shirt.

**Pull request merged**